nodejs/device/src/index.ts

```diff
@@ -0,0 +1,318 @@
+import { SymmetricKeySecurityClient } from 'azure-iot-security-symmetric-key';
+import { ProvisioningDeviceClient } from 'azure-iot-provisioning-device';
+import { Client, Message, Twin } from 'azure-iot-device';
+import { createHmac } from 'crypto';
+import { v4 as uuidv4 } from 'uuid';
+import { promisify } from 'util';
+
+const provisioningHost = process.env.THIEF_DEVICE_PROVISIONING_HOST;
+const idScope = process.env.THIEF_DEVICE_ID_SCOPE;
+const groupSymmetricKey = process.env.THIEF_DEVICE_GROUP_SYMMETRIC_KEY;
+const registrationId = process.env.THIEF_DEVICE_ID;
+const requestedServicePool = process.env.THIEF_REQUESTED_SERVICE_POOL;
+const runId = uuidv4();
+if (!provisioningHost || !idScope || !groupSymmetricKey || !registrationId || !requestedServicePool) {
+  throw new Error('Required environment variable is undefined.');
+}
+
+const settings: thiefSettings = {
+    thiefMaxRunDurationInSeconds: 0,
+    thiefPropertyUpdateIntervalInSeconds: 30,
+    thiefWatchdogFailureIntervalInSeconds: 300,
+    thiefAllowedClientLibraryExceptionCount: 10,
+    pairingRequestTimeoutIntervalInSeconds: 900,
+    pairingRequestSendIntervalInSeconds: 30,
+    sendMessageOperationsPerSecond: 1,
+    sendMessageAllowedFailureCount: 1000,
+    receiveC2dIntervalInSeconds: 20,
+    receiveC2dAllowedMissingMessageCount: 100,
+    reportedPropertiesUpdateIntervalInSeconds: 10,
+    reportedPropertiesAllowedFailureCount: 50,
+}
+
+class DeviceApp {
+  private async createDeviceClientUsingDpsGroupKey() {
+    const deviceKey = createHmac('SHA256', Buffer.from(groupSymmetricKey, 'base64')).update(registrationId, 'utf8').digest('base64');
+    const provisioningSecurityClient = new SymmetricKeySecurityClient(registrationId, deviceKey);
+    const provisioningClient = ProvisioningDeviceClient.create(provisioningHost, idScope, new ProvisioningTransport(), provisioningSecurityClient);
+    const registrationResultPromise = provisioningClient.register();
+    if (!registrationResultPromise) {
+      throw new Error('.register() did not return a promise when it should have.');
+    }
+    const registrationResult = await registrationResultPromise.catch(e => {throw new Error('DPS registration failed: ' + e.message)});
+    const connectionString = 'HostName=' + registrationResult.assignedHub + ';DeviceId=' + registrationResult.deviceId + ';SharedAccessKey=' + deviceKey;
+    const hubClient = Client.fromConnectionString(connectionString, IotHubTransport);
+    await hubClient.open().catch(e => {throw new Error('Opening client failed: ' + e.message)});
+    return hubClient;
+  }
+
+  private pairWithService() {
+    console.info('Starting pairing operation.');
+    const pairingRequestPatch = {
+      thief: {
+        pairing: {
+          requestedServicePool: requestedServicePool,
+          serviceInstanceId: null,
+          runId: runId
+        }
+      }
+    };
+
+    const pairWithServiceAttemptRecursive = () => {
```

Comment:
A function that returns a promise that is also recursive?  I kinda get this, but it hurts my head to try to reason about the behavior.  I understand what you're doing here, but I think you could just as easily do this iteratively, and I think it would be clearer.

Reply:
@YoDaMa @anthonyvercolano What are your thoughts on keeping this recursive vs redoing it iteratively? It's definitely possible to have a good iterative solution too, Bert came up with one.

Reply:
After looking at it more, I think I'll do it iteratively